Improvement: garden keybinds

Cedric-Ab opened the pull request:
<!-- remove all unused parts  -->

## What

You can now set 5 keybind layouts, each with different keybinds.
You can then for each crop select what layout you'd wish to use.
Also moved it into it's own category

<details>
<summary>Images</summary>

![image](https://github.com/user-attachments/assets/9106ed1a-685a-4f66-ae8b-f75ab902a791)


</details>

## Changelog Improvements
+ Improved garden keybinds. - Cédric Ab
    * Now set up to 5 different keybind layouts to choose from with each crop.

Cedric-Ab committed:
moved keybind settings to own category

Cedric-Ab committed:
do all java changes

do GUI and config part, now only kotlin side has to be done

Cedric-Ab committed:
add buttons in profiles

add disableAll and defaultAll

Cedric-Ab committed:
implemented logic

should work but of course doesn't

Cedric-Ab committed:
made it work

Cedric-Ab committed:
do the warnings

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/garden/farming/GardenCustomKeybinds.kt:
```diff
@@ -67,43 +76,96 @@ object GardenCustomKeybinds {
-                update()
-            }
+        var allKeybindings = listOf(
+            config.layout1.attack, config.layout1.useItem, config.layout1.left, config.layout1.right,
+            config.layout1.forward, config.layout1.back, config.layout1.jump, config.layout1.sneak,
+            config.layout2.attack, config.layout2.useItem, config.layout2.left, config.layout2.right,
+            config.layout2.forward, config.layout2.back, config.layout2.jump, config.layout2.sneak,
+            config.layout3.attack, config.layout3.useItem, config.layout3.left, config.layout3.right,
+            config.layout3.forward, config.layout3.back, config.layout3.jump, config.layout3.sneak,
+            config.layout4.attack, config.layout4.useItem, config.layout4.left, config.layout4.right,
+            config.layout4.forward, config.layout4.back, config.layout4.jump, config.layout4.sneak,
+            config.layout5.attack, config.layout5.useItem, config.layout5.left, config.layout5.right,
+            config.layout5.forward, config.layout5.back, config.layout5.jump, config.layout5.sneak
```
Small Suggestion:
Better to add a function that gets all properties as a list and then simply combine the 5 lists. Makes it better scale able.

Cedric-Ab replied:
@Thunderblade73 
would it be better to use stuff like getting all the variables in the config to then dynamically do the stuff. so you could just adjust the number of layouts in config and then the kotlin would automatically just work with the amount of layouts in the config?
My main concern with this is the unneeded complexity this would generate

Cedric-Ab replied:
@Thunderblade73 
like the new version?

Cedric-Ab committed:
Merge branch 'hannibal002:beta' into improvement-garden-keybinds-2

Cedric-Ab committed:
Update GardenCustomKeybinds.kt

fix small mistakes
+ redo getting allKeybindings in onConfigLoad

Cedric-Ab commented:
@Thunderblade73
would it be better to use stuff like getting all the variables in the config to then dynamically do the stuff. so you could just adjust the number of layouts in config and then the kotlin would automatically just work with the amount of layouts in the config?
My main concern with this is the unneeded complexity this would generate

Cedric-Ab committed:
Update GardenCustomKeybinds.kt

fix "detekt failure"

Thunderblade73 commented:
> @Thunderblade73 would it be better to use stuff like getting all the variables in the config to then dynamically do the stuff. so you could just adjust the number of layouts in config and then the kotlin would automatically just work with the amount of layouts in the config? My main concern with this is the unneeded complexity this would generate

Yeah that would be better imo. If done correctly it should reduce complexity and not add more (as far as I can tell).
Also if you want to give the layout enum more functionallity I would suggest to move it to kotlin, since java code gets more cluttered and the config supports kotlin enums.

Btw. enums have a way to get all entries ```.values``` (for java) and ```.entries``` for kotlin (before the dot is the Enum Type)

Cedric-Ab committed:
fix it not starting on IslandChange

add workaround for this, as GardenAPI should be calling GardenToolChangeEvent on island change

Cedric-Ab commented:
non essential TODOs:
fix detekt thing
make kotlin detect amount of layouts

Cedric-Ab commented:
> > @Thunderblade73 would it be better to use stuff like getting all the variables in the config to then dynamically do the stuff. so you could just adjust the number of layouts in config and then the kotlin would automatically just work with the amount of layouts in the config? My main concern with this is the unneeded complexity this would generate
> 
> Yeah that would be better imo. If done correctly it should reduce complexity and not add more (as far as I can tell). Also if you want to give the layout enum more functionallity I would suggest to move it to kotlin, since java code gets more cluttered and the config supports kotlin enums.
> 
> Btw. enums have a way to get all entries `.values` (for java) and `.entries` for kotlin (before the dot is the Enum Type)

do you think I could replace the map: layouts with the enum?
and would it need to be in kotlin or could it stay in java?

Thunderblade73 commented:
> do you think I could replace the map: layouts with the enum? and would it need to be in kotlin or could it stay in java?

Yes, you can replace the map.
The can stay in java yes, but makeing them in kotlin is easier.

Cedric-Ab committed:
start implementing kontlin enum

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/garden/farming/keybinds/KeyBindLayouts.kt:
```diff
@@ -0,0 +1,31 @@
+                getAllKeybindingsFromLayout(layout)
+            ).forEach { (keyBinding, setKeyProperty) ->
+                map[keyBinding] = setKeyProperty.get() // Update map directly
+            }
```
Imo a ``var map: Map`` is better than a ``val map: MutableMap`` here.
Also:
```suggestion
            map = keyBindings.associate(
                getAllKeybindingsFromLayout(layout).get()
            )
```

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/garden/farming/keybinds/KeyBindLayouts.kt:
```diff
@@ -0,0 +1,31 @@
+            val keyBindings = listOf(
+                keyBindAttack, keyBindUseItem, keyBindLeft, keyBindRight,
+                keyBindForward, keyBindBack, keyBindJump, keyBindSneak
+            )
```
The ``with`` only applies is used for this so close it here.
Something like that.
```Kotlin
            val keyBindings = with(mcSettings) {
             listOf(
                keyBindAttack, keyBindUseItem, keyBindLeft, keyBindRight,
                keyBindForward, keyBindBack, keyBindJump, keyBindSneak
            )
            }
```

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/garden/farming/GardenCustomKeybinds.kt:
```diff
@@ -67,43 +79,106 @@ object GardenCustomKeybinds {
+//         }
+//     }
+
+    @JvmStatic
+    fun getAllKeybindingsFromLayout(layout: KeyBindLayout) = listOf(
+        layout.attack, layout.useItem, layout.left, layout.right,
+        layout.forward, layout.back, layout.jump, layout.sneak
+    )
```
An extension function works better here. (Or make it a member function in java, with an ArrayList/Array)
```suggestion
    @JvmStatic
    fun KeyBindLayout.getAllKeybindings() = listOf(
        this.attack, this.useItem, this.left, this.right,
        this.forward, this.back, this.jump, this.sneak
    )
```

Cedric-Ab replied:
done

Cedric-Ab committed:
Update KeyBindConfig.java

remove unused enum Layouts

Cedric-Ab committed:
Update KeyBindLayout.java

add getAllKeybindings()
+ formatting

Cedric-Ab committed:
finish enum KeyBindLayouts

also adapt to the enum

Cedric-Ab committed:
Update enum KeyBindLayouts

make map non mutable and change how it's built

Cedric-Ab committed:
Update ConditionalUtils.kt

add onToggle that accepts a collection

Cedric-Ab committed:
Update GardenCustomKeybinds.kt

add @Suppress("unused")  on events

Cedric-Ab committed:
Update KeyBindLayouts.kt

fix detekt failures

Cedric-Ab committed:
Update ConditionalUtils.kt

remove spread operator for detekt

Cedric-Ab committed:
Merge branch 'hannibal002:beta' into improvement-garden-keybinds-2

Cedric-Ab commented:
@Thunderblade73 
I hope it's good now

Cedric-Ab committed:
remove lastCrop in GardenCustomKeybinds.kt

Cedric-Ab committed:
Update KeyBindLayouts.kt

fix wrong name displaying in selection

Cedric-Ab committed:
update on layout selection change

Cedric-Ab committed:
move GardenCustomKeybinds.kt into keybinds

Cedric-Ab committed:
Revert "move GardenCustomKeybinds.kt into keybinds"

This reverts commit 2995018087b874c39857e4e6b9841a7e6ffa74fe.

Cedric-Ab commented:
TODO: move GardenCustomKeybinds.kt into keybinds

Cedric-Ab commented:
@Thunderblade73 
Is cropLayoutSelection initialized on startup?

Thunderblade73 commented:
> @Thunderblade73 Is cropLayoutSelection initialized on startup? is configloadevent called on startup?

It is loaded via Gson at ConfigLoadEvent (or what ever it is called) or if no value is present in the config json than it calls the constructor.

Cedric-Ab commented:
> > @Thunderblade73 Is cropLayoutSelection initialized on startup? is configloadevent called on startup?
> 
> It is loaded via Gson at ConfigLoadEvent (or what ever it is called) or if no value is present in the config json than it calls the constructor.

The cropLayoutSelection is a map inside of the kotlin, you got that like this, right?

Thunderblade73 commented:
> The cropLayoutSelection is a map inside of the kotlin, you got that like this, right?

Nope, I saw the value in KeyBindConfig and assume you where referencing it,

Cedric-Ab commented:
> > The cropLayoutSelection is a map inside of the kotlin, you got that like this, right?
> 
> Nope, I saw the value in KeyBindConfig and assume you where referencing it,

So is ConfigLoadEvent called upon launch?

Thunderblade73 commented:
> is configloadevent called on startup?

yes

> Is cropLayoutSelection initialized on startup?

Yes, but you override it every update().
Btw. it does not need to exist since you can simply add a extension function to CropType that gets the value from an when since it does not compute anything since the getLayoutByDisplayName(something.toString()) is redundant because ``something`` is already a Layout

Cedric-Ab commented:
> > is configloadevent called on startup?
> 
> yes
> 
> > Is cropLayoutSelection initialized on startup?
> 
> Yes, but you override it every update(). Btw. it does not need to exist since you can simply add a extension function to CropType that gets the value from an when since it does not compute anything since the getLayoutByDisplayName(something.toString()) is redundant because `something` is already a Layout

Yes, I see
This is a thing that needs to be done

### TODO

Cedric-Ab committed:
remove unnecessairy method

KeyBindLayouts.getLayoutByDisplayName(KeyBindLayouts.to_String())
is redundant lol

Cedric-Ab committed:
Update GardenCustomKeybinds.kt

add extension function to CropType and remove unneeded stuff

Cedric-Ab committed:
remove @Suppress("unused")

Cedric-Ab committed:
definitively remove IslandChangeEvent workaround

CalMWolfs committed:
Merge branch 'beta' into improvement-garden-keybinds-2

Cedric-Ab committed:
implement #3186

Cedric-Ab committed:
Merge branch 'beta' into improvement-garden-keybinds-2

Cedric-Ab committed:
Detekt

Cedric-Ab commented:
TODO:

- [ ] Move GardenCustomKeybinds.kt into keybinds

Cedric-Ab committed:
Merge branch 'beta' into improvement-garden-keybinds-2

Cedric-Ab committed:
fix shit

Cedric-Ab commented:
so when should this get merged?

Cedric-Ab committed:
Merge branch 'hannibal002:beta' into improvement-garden-keybinds-2

Cedric-Ab committed:
add ConfigFix

Cedric-Ab committed:
Merge branch 'hannibal002:beta' into improvement-garden-keybinds-2

CalMWolfs committed:
Merge branch 'beta' into improvement-garden-keybinds-2

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/Cedric-Ab/improvement-garden-keybinds-2

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/features/garden/farming/GardenCustomKeybinds.kt

Cedric-Ab committed:
Merge branch 'hannibal002:beta' into improvement-garden-keybinds-2

Cedric-Ab committed:
Merge branch 'hannibal002:beta' into improvement-garden-keybinds-2

Cedric-Ab committed:
Merge branch 'hannibal002:beta' into improvement-garden-keybinds-2
